src/Voron/Impl/Scratch/ScratchBufferPool.cs

```diff
@@ -129,30 +129,45 @@ public PageFromScratchBuffer Allocate(LowLevelTransaction tx, int numberOfPages)
 
             if (sizeAfterAllocation >= (_sizeLimit * 3) / 4 && oldestActiveTransaction > current.OldestTransactionWhenFlushWasForced)
             {
-                // we may get recursive flushing, so we want to avoid it
-                if (tx.Environment.Journal.Applicator.IsCurrentThreadInFlushOperation == false)
-                {
-                    // We are starting to force a flush to free scratch pages. We are doing it at this point (80% of the max scratch size)
-                    // to make sure that next transactions will be able to allocate pages that we are going to free in the current transaction.
-                    // Important notice: all pages freed by this run will get ValidAfterTransactionId == tx.Id (so only next ones can use it)
+                // because we already know that we have a lazy transaction, we'll skip forcing a flush here, in favor of setting the
+                // transaction to be non lazy, which will trigger a flush after this is over. This will also release the pending read
+                // transaction, which can run in parallel to the next transactions, and allow us to free up scratch space without stalling
+                // the current write transaction until the entire scratch file (which can be very large) is flushed
+
+                // if the transaction is lazy, then there is a long running lazy transaction that is being held to prevent us 
+                // from flushing from the scratch buffer to the data file. In this case, setting this transaction as non lazy
+                // will force us to actually flush everything when this transaction is committed. This will allow us to flush
+                // the data to the data file before we run out of space in the scratch file
```

Comment:
How do we already have the flushing lock? In the allocate? 
At that point we can dispose of the read tx, yes, but only if we first flush the journal.

We tried disabling lazy transactions when the scratch file is full, I think we still run out of space.